src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -763,54 +842,222 @@ cbuf_unmap_at(spdid_t s_spd, unsigned int cbid, spdid_t d_spd, vaddr_t d_addr)
-	int ret = 0;
+	struct cbuf_info *cbi;
+	unsigned int sz;
+	spdid_t spdid = f_spd;
+	vaddr_t dest;
+	int ret = -1;
+	int f_cbid;
+	int flags;
+	struct cbuf_info *f_cbi;
+	void *memcpy_ret;
+	struct cbuf_maps *c;	/* c for current because we love one letter names */
+
+	cbi        = cmap_lookup(&cbufs, s_cbid);
+	if (!cbi) BUG();
+	sz = cbi->size;
+	printc("forking cbuf %d from spdid %d to spdid %d, with size %x\n", s_cbid, o_spd, f_spd, sz);
+
+	/* cmap_lookup returns original owner so need to recurse through mappings until we find the one with spdid O */
+	c = &cbi->owner;
+	while (c->spdid != o_spd) c = c->next; // what about infinite loops???
+	assert(c->spdid == o_spd);
+	dest = c->addr;
+
+	/* create cbuf */
+	f_cbid = __cbuf_create(spdid, sz, 0, dest);
+	if (f_cbid < 0) { // only do these steps if initial create failed because of lack of map
+		if (!__cbuf_register_map_at(o_spd, f_spd, cbi->cbid, f_cbid * -1)) { 
+			//goto done; 
+		}
+		f_cbid = __cbuf_create(spdid, sz, f_cbid * -1, dest);
+	}
 
-	printl("cbuf_fork_spd\n");
+	if (f_cbid <= 0) { 
+		printd("cbi didn't agree\n");
+		goto done;
+	}
+	assert(f_cbid);
+
+	/* copy into cbuf */
+	assert(cbi->mem);
+	f_cbi = cmap_lookup(&cbufs, f_cbid);
+	if (!f_cbi) BUG(); 				/* we literally just made this in cbuf_create so if it's not there, something is very wrong */
+	assert(f_cbi->mem);
+	memcpy_ret = memcpy(f_cbi->mem, cbi->mem, sz);
+	assert(memcpy_ret == f_cbi->mem);
+
+	/* Do a sanity check and REMOVE THIS once we're kinda confident stuff works */
+	if (memcmp(f_cbi->mem, cbi->mem, sz)) { printc("cbufs do not actually match. Womp womp :(\n"); BUG(); }
+	
+	ret = 0;
 
-	CBUF_TAKE();
-	s_cci = cbuf_comp_info_get(s_spd);
-	if (unlikely(!s_cci)) goto done;
-	d_cci = cbuf_comp_info_get(d_spd);
-	/* FIXME: This should be making copies to avoid sharing */
-	__cbuf_copy_cci(s_cci, d_cci);
+	if (copy_cinfo) {
+		printc("About to try mapping cinfo cbuf into Q\n");
+		vaddr_t q_daddr = (vaddr_t)valloc_alloc(cos_spd_id(), 11, sz/PAGE_SIZE);
+		if (unlikely(!dest)) return -1;
+		flags = MAPPING_RW;
+		flags |= MAPPING_NO_VALLOC;
+		if (q_daddr != __cbuf_map_at(f_spd, f_cbid, 11, q_daddr | flags)) return -1;
+		ret = q_daddr;
+	}
 
 done:
-	CBUF_RELEASE();
 	return ret;
+
+}
+
+/*
+ * This is meant to be silly code to help me (Teo) debug LockdownOS but maybe other people would like it too.
+ * Prints out info relevant to component cci. Does NOT modify stuff.
+ */
+static void
+__get_nfo(struct cbuf_comp_info *cci)
+{
+	printc("Now iterating over global cbuf list\n");
+	// iterate over cbi_head until done
+	struct cbuf_info *current = cbi_head;
+	if (!current) {
+		printc("There should be at least one cbuf in the system\n");
+		BUG();
+	}
+
+	do {
+		// The cbuf_info list we made is actually crap because it copies the pointers so if the structure is updated it is no longer valid (how to fix???)
+		// instead, re-get the cbi from the cbuf_id which we assume is constant
```

Comment:
Sigh... I was hoping to figure out a smarter way to do this in general and then remove the comment altogether but clearly that didn't happen.